compiler/src/dotty/tools/dotc/transform/sjs/ExplicitJSClasses.scala

```diff
@@ -12,54 +17,728 @@ import core.Types._
+          if (isExposedModule(declSym)) {
+            val getter = cls.info.decls.lookup(jsobjectGetterNameFor(declSym)).asTerm
+            newDecls += DefDef(getter, ref(declSym))
+          }
+        }
+
+        newDecls += decl
+      }
+
+      cpy.Template(tree)(tree.constr, fixedParents, Nil, tree.self, newDecls.result())
+    }
+  }
+
+  // This method, together with transformTypeDef, implements step (C)
+  override def prepareForTypeDef(tree: TypeDef)(using Context): Context = {
+    val sym = tree.symbol
+    if (sym.isClass && isLocalJSClass(sym)) {
+      val jsclassValName = LocalJSClassValueName.fresh(sym.name.toTermName)
+      val jsclassVal = newSymbol(ctx.owner, jsclassValName, EmptyFlags, defn.AnyRefType, coord = tree.span)
+      val state = myState
+      state.localClass2jsclassVal(sym) = jsclassVal
+      state.notYetSelfReferencingLocalClasses += sym
+    }
+    ctx
+  }
+
+  // This method, together with prepareForTypeDef, implements step (C)
+  override def transformTypeDef(tree: TypeDef)(using Context): Tree = {
+    val sym = tree.symbol
+    if (sym.isClass && isLocalJSClass(sym)) {
+      val state = myState
+      val cls = sym.asClass
+
+      val rhs = {
+        val typeRef = tree.tpe
+        val clazzValue = clsOf(typeRef)
+        val superClassCtor = genJSConstructorOf(tree, extractSuperTpeFromImpl(tree.rhs.asInstanceOf[Template]))
+        val fakeNewInstances = {
+          /* We need to use `reverse` because the Scope returns elements in reverse order compared to tree definitions.
+           * The back-end needs the fake News to be in the same order as the corresponding tree definitions.
+           */
+          val ctors = cls.info.decls.lookupAll(nme.CONSTRUCTOR).toList.reverse
+          val elems = ctors.map(ctor => fakeNew(cls, ctor.asTerm))
+          JavaSeqLiteral(elems, TypeTree(defn.AnyRefType))
+        }
+        ref(jsdefn.Runtime_createLocalJSClass).appliedTo(clazzValue, superClassCtor, fakeNewInstances)
+      }
+
+      val jsclassVal = state.localClass2jsclassVal(sym)
+      if (state.notYetSelfReferencingLocalClasses.remove(cls)) {
+        Thicket(List(tree, ValDef(jsclassVal, rhs)))
+      } else {
+        /* We are using `jsclassVal` inside the definition of the class.
+         * We need to declare it as var before and initialize it after the class definition.
+         */
+        jsclassVal.setFlag(Mutable)
+        Thicket(List(
+            ValDef(jsclassVal, Literal(Constant(null))),
+            tree,
+            Assign(ref(jsclassVal), rhs)
+        ))
+      }
+    } else {
+      tree
+    }
+  }
+
+  /** Creates a fake invocation of the the given class with the given constructor. */
+  def fakeNew(cls: ClassSymbol, ctor: TermSymbol)(using Context): Tree = {
+    /* TODO This is not entirely good enough, as it break -Ycheck for generic
+     * classes. Erasure restores the consistency of the fake invocations.
+     * Improving this is left for later.
```

Comment:
@smarter Any idea how to improve this?

My best idea so far is to leave holes at this phase, and add *another* phase between Erasure and LambdaLift that would create the fake New nodes. Indeed, after erasure, the problem of type parameters does not exist anymore, and we wouldn't break consistency.

Reply:
I think the best option would be to tweak LambdaLift to make it do what you want. But failing that, I agree that adding the new after Erasure would be nicer than doing it here indeed.

Reply:
Regarding making lambdalift do what I want, I've been trying to come up with exactly what I want. And in fact I cannot think of a better way to receive the information I want than the one I have now (at least for ways that maintain semantic meaning of the Trees we have in the compiler at the phases we have).

So the question remains how to produce that information, and AFAICT modifying lambdalift will not be easier than having a miniphase create those fake New nodes, so I'm more inclined to go that route.